### ISSUE TITLE
PrePARE missing  _FillValue exit

### ISSUE THREAD

**Issue by wachsylon** [PCMDI/cmor#367]
Hi,
I have data files with variables which do not include _FillValue and _MissingValue attributes. This causes the following error wihtin PrePARE:
```
Traceback (most recent call last):
...
  File "/work/bm0021/miniconda2/envs/cmor/lib/python2.7/site-packages/cmip6_cv/PrePARE/PrePARE.py", line 446, in ControlVocab
    varmissing = infile.variables[var[0]]._FillValue[0]
AttributeError: _FillValue
```
However, I would like to have an overview about all missing or wrong variable attributes so PrePARE should not exit at this point without checking the other ones. I installed it via conda.
Is there already an option to ignore that error or is it a bug?
I have not looked into the src of PrePARE yet.
Best regards,
Fabi

**Comment by taylor13** [comment]
As you know, PrePARE is tailored to examine files that are to be published on ESGF with metadata conforming to the CMIP6 specifications.  For CMIP6 the variable attributes missing_value and FillValue are required, but I'm surprised that PrePARE  checks that they are there, but if it does this is desirable for CMIP6.  If you look at the code, let us know what you find.
thanks,
Karl

**Comment by wachsylon** [comment]
Just to be clear: I think it would be good for data providers - who want to publish data via ESGF - to know _all_ attributes that are missing in the files instead of only `_FillValue`.

I saw that the code causing this error was changed since commit 
6ce9ee2320e82c72a93ab15dd3993f401201bb92
so I will once more check whether I used an old version however I am pretty sure that I used the newest Conda. If I can install PrePare with the GitHub code it would be great if you can provide a short guidance about which python and so on..

**Comment by wachsylon** [comment]
And when I include the missing file from the github repo:
`cp cmor-git/LibCV/PrePARE/out_names_tests.json /mnt/lustre01/work/bm0021/miniconda2/envs/cmor/lib/python2.7/site-packages/cmip6_cv/PrePARE/out_names_tests.json`
I receive again:
```
Traceback (most recent call last):
  File "/work/bm0021/miniconda2/envs/cmor/bin/PrePARE", line 11, in <module>
    load_entry_point('CMOR==3.3.3', 'console_scripts', 'PrePARE')()
  File "/work/bm0021/miniconda2/envs/cmor/lib/python2.7/site-packages/cmip6_cv/PrePARE/PrePARE.py", line 857, in main
    errors += sequential_process(source)
  File "/work/bm0021/miniconda2/envs/cmor/lib/python2.7/site-packages/cmip6_cv/PrePARE/PrePARE.py", line 710, in sequential_process
    checker.ControlVocab(ncfile)
  File "/work/bm0021/miniconda2/envs/cmor/lib/python2.7/site-packages/cmip6_cv/PrePARE/PrePARE.py", line 495, in ControlVocab
    self.dictVar['_FillValue'][0],
KeyError: '_FillValue'
```
which corresponds to the previous error.

**Comment by taylor13** [comment]
The request is that PrePARE should not error exit on encountering the first error.  Rather it should continue on and complete all its checks.  This will help users to more quickly modify their output files to be compliant with the specs.

**Comment by mauzey1** [comment]
@taylor13 @doutriaux1 Getting back to this, it looks like it tries to get the values for the attributes `_FillValue` and `units` from the file before it checks all attributes.
https://github.com/PCMDI/cmor/blob/2e9f6b2f6acbd25d15b6c46a26176632c864ac8b/LibCV/PrePARE/PrePARE.py#L536-L542
Why do we need `_FillValue` and `units` from the file if we are checking them against the attributes in the table?

**Comment by doutriaux1** [comment]
@mauzey1 I'm not positive but it looks like we are trying to validate that values stored in the file match what they should be? @taylor13 any idea?

**Comment by mauzey1** [comment]
@doutriaux1 The cmip6_cv.setup_variable line is setting up a variable that has attribute values from the table.  However, it is getting `_FillValue` from the file and using that as the value for the table's `_FillValue` and `missing_value`.  It doesn't change the table value of `units` though.

**Comment by taylor13** [comment]
As we discussed, the suggestion is to either avoid passing any information about the file's units and _FillValue as arguments of cmip6_cv.setup_variable  (i.e., either remove the arguments self.dictVar['units'], self.dictVar['_FillValue'], or set them to some value serving as a flag that tells cmip6_cv.setup_variable to skip any code where these arguments are currently required).  There should be no need to read from the file _FillValue and units prior to checking them against the information in the table.

**Comment by mauzey1** [comment]
@taylor13 The variable `basin` in table `Ofx` is the only variable I have found that has the type integer.  
https://github.com/PCMDI/cmor/blob/2e9f6b2f6acbd25d15b6c46a26176632c864ac8b/TestTables/CMIP6_Ofx.json#L35-L54
I thought it would use -999 as its missing value.
https://github.com/PCMDI/cmor/blob/2e9f6b2f6acbd25d15b6c46a26176632c864ac8b/TestTables/CMIP6_Ofx.json#L9
However, when I got a NetCDF file for `basin` from ESGF I found out that it was using 0 as its `missing_value` and `_FillValue`.

Should we treat `basin` as a special case?  It's the only variable that has the attributes `flag_values` and `flag_meanings`.

**Comment by taylor13** [comment]
"0" should definitely *not* be declared as `missing_value` or `_FillValue`.  "0" indicates the region is "global_land", according to `flag_meanings` and `flag_values`.  Where is CMOR getting missing_value="0"?  I suspect that somewhere along the line -999 gets transformed to float or something, and then gets transformed back to 0.  The `missing_value` and `_FillValue` attributes in the file should be as specified in the table: -999.

I don't think this should be a special case (other than it is declared integer rather than float).

**Comment by mauzey1** [comment]
@doutriaux1 @taylor13 

I submitted a pull request that will make PrePARE  get the units and missing values from the table instead of the NetCDF file.  I also added a parameter for the missing integer value when setting up the variable.

Alternatively, we could just get rid of the units and missing value parameters from the variable setting functions since cmor_CV_variable can get the values directly from the table.  I don't think anything other than PrePARE uses these functions.
https://github.com/PCMDI/cmor/blob/84fd5213bf9fadd93beb228d689c3e0b825e25ed/Src/cmor_CV.c#L2292-L2295
